### PR TITLE
feat: refreshInterval as a function

### LINKED ISSUE #1683: refreshInterval as function
### Discussed in https://github.com/vercel/swr/discussions/1630

<div type='discussions-op-text'>

<sup>Originally posted by **correttojs** November 11, 2021</sup>
currently the `refreshInterval` option can only be a number. In some cases this number should be dynamic, dependent on the updated fetched data.
For example, the fetcher returns a list of items and a value corresponding to the next expected update. </div>

### PR DESCRIPTION
**Pull request by Andrewnt219**
## Issue
Closes #1683 

## Description

In addition to passing a number, `refreshInterval` can also receive a function that returns a number as an interval. The function has access to the updated data.

Do I need to add anything to the examples folder? I'd like to update the docs as well, but it's in a different repo.

## Test plan

- Function can be passed to `refreshInterval`
- Data passed to `refreshInterval`  is updated data

### CODE REVIEW

File: src/use-swr.ts [hunk 1]
```diff
@@ -444,7 +450,12 @@ export const useSWRHandler = <Data = any, Error = any>(
       // - `force` is true, which is the start of polling
       // - or `timer` is not 0, which means the effect wasn't canceled
       if (refreshInterval && timer !== -1) {
-        timer = setTimeout(execute, refreshInterval)
+        // Use the passed interval
+        // ...or invoke the function with the updated data to get the interval
+        const interval = isFunction(refreshInterval)
+          ? refreshInterval(data)
+          : refreshInterval
+        timer = setTimeout(execute, interval)
```
Comment (shuding) [inline src/use-swr.ts:458]: If `interval` is `0` we don't want to continue polling:

```suggestion
        if (interval) {
          timer = setTimeout(execute, interval)
        }
```

Comment (Andrewnt219) [reply src/use-swr.ts:458]: Ah that's right. I should move the logic outside of the current if to prevent the double check.

**Comment by Andrewnt219** [comment]
Should I add a test for function returns 0?

**Comment by Andrewnt219** [comment]
There's a flaw in the test for updated data, I'll update it soon.

File: src/use-swr.ts (hunk 1 above)
Comment (shuding) [reply src/use-swr.ts:459]: Yep that would be great!

**Review by shuding — APPROVED** [review]
Perfect, thank you!

**Comment by shuding** [comment]
I think we can add a test for the case of returning `0` (can be another PR of course), this is already looking good to me. Thanks a lot for your contribution!

**Comment by Andrewnt219** [comment]
No problem, I have used swr a lot and it feels great to give back. I'll have a follow up PR for the test soon. 

Do I have to update somewhere in the docs site as well?

**Comment by shuding** [comment]
Thank you! You can update the description for that option here: https://github.com/vercel/swr-site/blob/master/pages/docs/options.en-US.mdx (and update other languages in English too, other contributors can pick up the translation work) 🎉

**Pull request merged**